controller_manager/src/ros2_control_node.cpp

```diff
@@ -0,0 +1,61 @@
+// Copyright 2020 ROS2-Control Development Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+
+#include "controller_manager/controller_manager.hpp"
+#include "hardware_interface/robot_hardware.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "test_robot_hardware/test_robot_hardware.hpp"
+
+using namespace std::chrono_literals;
+
+constexpr const auto kLoggerName = "ros2_control_node";
+
+int main(int argc, char ** argv)
+{
+  rclcpp::init(argc, argv);
+
+  std::shared_ptr<rclcpp::Executor> executor =
+    std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
+  std::string manager_node_name = "controller_manager";
+  rclcpp::TimerBase::SharedPtr timer;
+
+  auto cm = std::make_shared<controller_manager::ControllerManager>(
+    // TODO(anyone): remove robot_hw when ResourceManager is added
+    // since RobotHW is not a plugin we had to take some type of robot
+    std::make_shared<test_robot_hardware::TestRobotHardware>(),
+    executor,
+    manager_node_name);
+
+  // Declare default controller manager rate of 100Hz
+  cm->declare_parameter("update_time_ms", 10);
+  // load controller_manager update time parameter
+  int update_time_ms;
+  if (!cm->get_parameter("update_time_ms", update_time_ms)) {
+    throw std::runtime_error("update_time parameter not existing or empty");
+  }
+  RCLCPP_INFO(rclcpp::get_logger(kLoggerName), "update time is %.3f ms", update_time_ms);
+
+  timer = cm->create_wall_timer(
+    std::chrono::milliseconds(update_time_ms),
+    std::bind(&controller_manager::ControllerManager::update, cm.get()),
+    cm->deterministic_callback_group_);
```

Comment:
this looks a bit like a dependency inversion to me. If the executor is handled outside of the controller manager in its own node, then the associated callback groups should be handled externally as well. 
With this approach - essentially having the controller manager's callback groups exposed publicly - every node can temper with the execution model of the controller manager and provoke a undesired behavior given the assumptions within the CM.

Reply:
Realistically, I see two executors being used here. The first one spawns the controller and handles the update (as well as the read and write) loop with a given frequency. That executor can then be assigned a higher realtime priority by the operating system to guarantee a deterministic behavior.
We have to introduce a second node which handles the async callbacks to low prio calls such as list controllers etc. 

That workaround is needed as we're targeting foxy and we can't leverage the changes introduced in here: https://github.com/ros2/rclcpp/pull/1218

Reply:
Please submit a PR with a suggested alternative

Reply:
@Karsten1987 I agree with you. I added here are @v-lopez  and @gavanderhoorn commented that they want to temper with the execution of update function.

Or should they then just call specialize `ControllerManager` in a new class?